### ISSUE TITLE
Validation acc varies in different test batch size.

### ISSUE THREAD

**Issue by Bob130** [bearpaw/pytorch-pose#53]
@bearpaw thank you for your wonderful work,
I have trained my network. While, when I validated my best model in different test batch size, the validation acc suffered from different results, as I  think the acc should be independent of test batch size.

test batchsize------Val Acc
1--------------------0.8743
6--------------------0.8660
16-------------------0.8685

And what's the test batch size do you use in the result you published？

**Comment by bearpaw** [comment]
Hi @Bob130,

Did you test with the evaluation mode?

**Comment by djangogo** [comment]
Hi @Bob130 
I think this may be related with BatchNorm, because when it is in eval model it will produce the mean and variance according to current batch data.


[Simon Wang](https://discuss.pytorch.org/t/performance-highly-degraded-when-eval-is-activated-in-the-test-phase/3323/19?u=hitictoc)
> Exactly what I said above: with a small batch size, running average estimate is unstable. So you get worse results using that.

**Comment by Bob130** [comment]
> Hi @Bob130,
> 
> Did you test with the evaluation mode?

@bearpaw ,
Yes, I validated my best model with evaluation mode. The command is shown as below with different test batch size.
`CUDA_VISIBLE_DEVICES=1  python  example/mpii.py -a hg --stacks 8 --blocks 1 --checkpoint checkpoint/mpii_pytorch-0.1/s8b1  -j 10   --test-batch 16   --resume  checkpoint/mpii_pytorch-0.1/s8b1/model_best2.pth.tar -e`

**Comment by Bob130** [comment]
> Hi @Bob130
> I think this may be related with BatchNorm, because when it is in eval mode it will produce the mean and variance according to current batch data.
> 
> [Simon Wang](https://discuss.pytorch.org/t/performance-highly-degraded-when-eval-is-activated-in-the-test-phase/3323/19?u=hitictoc)
> 
> > Exactly what I said above: with a small batch size, running average estimate is unstable. So you get worse results using that.

@djangogo ,
Maybe. I tried with more test batch size and got the result below.
![image](https://user-images.githubusercontent.com/29658946/45539687-0cad4d00-b83d-11e8-9ca7-967f6ac216dd.png)
When batch size increased from 6 to 64, the Val Acc got better. Which confused me is that Val Acc increased nearly 0.01 using batch size 1 compared with batch size 6.

**Comment by John6333** [comment]
Hi @Bob130, 

I think the reason validation accuracy varies with batch size is that the accuracy used here is not "correctly defined", not related to Batch Norm layers.

According to the code, the final accuracy is the average of all the accuracies of different joints. But the number of annotated joints can be different from joint to joint, making the summation of all the accuracies on different joints improper.

What's more, the number of annotated joints can vary from batch to batch, so the summation of all the accuracies in different batches is also improper.
 
In my understanding,  the correct way to calculate accuracy is first to sum the number of correct predictions, then sum the number of annotated joints, then
accuracy = the number correct predictions / the number annotated joints.

**Comment by bearpaw** [comment]
@John6333 Thanks. 

@Bob130 Please run https://github.com/bearpaw/pytorch-pose/blob/master/evaluation/eval_PCKh.py or https://github.com/bearpaw/pytorch-pose/blob/master/evaluation/eval_PCKh.m to get the final accuracy (official definition of PCKh with original image scale).

**Comment by FantDing** [comment]
I think maybe you calculate the accuracy incorrectly. `avg_accuracy=total_correct_samples/total_samples`; But not `avg_accuracy=(acc1+acc2+...+accN)/N`
